tests/unit-tests/gateways/paypal/request.php

```diff
@@ -0,0 +1,261 @@
+
+	protected $products;
+
+	/**
+	 * Create $product_count simple products and store them in $this->products.
+	 *
+	 * @param int $product_count Number of products to create.
+	 */
+	protected function create_products( $product_count = 30 ) {
+		$this->products = array();
+		for ( $i = 0; $i < $product_count; $i++ ) {
+			$product = WC_Helper_Product::create_simple_product();
+			$product->set_name( 'Dummy Product ' . $i );
+			$this->products[] = $product;
+
+		}
+
+		// To test limit length properly, we need a product with a name that is shorter than 127 chars,
+		// but longer than 127 chars when URL-encoded.
+		if ( array_key_exists( 0, $this->products ) ) {
+			$this->products[0]->set_name( 'Dummy Product 😎😎😎😎😎😎😎😎😎😎😎' );
+		}
+
+	}
+
+	/**
+	 * Add products from $this->products to $order as items, clearing existing order items.
+	 *
+	 * @param WC_Order $order  Order to which the products should be added.
+	 */
+	protected function add_products_to_order( $order ) {
+		// Remove previous items.
+		foreach ( $order->get_items() as $item ) {
+			$order->remove_item( $item->get_id() );
+		}
+
+		// Add new products.
+		foreach ( $this->products as $product ) {
+			$item = new WC_Order_Item_Product();
+			$item->set_props( array(
+				'product'  => $product,
+				'quantity' => 3,
+				'subtotal' => wc_get_price_excluding_tax( $product, array( 'qty' => 3 ) ),
+				'total'    => wc_get_price_excluding_tax( $product, array( 'qty' => 3 ) ),
+			) );
+
+			$item->save();
+			$order->add_item( $item );
+		}
+
+	}
+
+	/**
+	 * Initialize the Paypal gateway and Request objects.
+	 */
+	public function setUp() {
+		parent::setUp();
+
+		$bootstrap = WC_Unit_Tests_Bootstrap::instance();
+		include_once $bootstrap->plugin_dir . '/includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php';
+
+		$this->paypal_gateway = new WC_Gateway_Paypal();
+		$this->paypal_request = new WC_Gateway_Paypal_Request( $this->paypal_gateway );
+	}
+
+
+	/**
+	 * Create Paypal request URL for $product_count number of products.
+	 *
+	 * @param int  $product_count Number of products to include in the order.
+	 * @param bool $testmode      Whether to test using sandbox or not.
+	 *
+	 * @return string
+	 * @throws WC_Data_Exception
+	 */
+	protected function get_request_url( $product_count, $testmode ) {
+		// Create products.
+		$this->create_products( $product_count );
+
+		$this->order = WC_Helper_Order::create_order( $this->user );
+		$this->add_products_to_order( $this->order );
+
+		// Set payment method to Paypal.
+		$payment_gateways = WC()->payment_gateways->payment_gateways();
+		$this->order->set_payment_method( $payment_gateways['paypal'] );
+		$this->order->calculate_totals();
+		$this->order->calculate_shipping();
+
+		return $this->paypal_request->get_request_url( $this->order, $testmode );
+	}
+
+	/**
+	 * Clean up order, deletes all products in order, too.
+	 */
+	protected function clean_up() {
```

Comment:
I think you can add a teardown method rather than calling this manually.

Reply:
`tearDown` is only called once at the end of each test method. Since all of the work in this is done in one test method, it wouldn't work.